File: solutions/2586-longest-square-streak-in-an-array/longest-square-streak-in-an-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestSquareStreak = function(nums) {
    // Convert nums to a sorted set to remove duplicates and have ordered numbers
    const numSet = new Set(nums);
    
    // Track the maximum streak length found
    let maxLength = 0;
    
    // Iterate through each number in the set
    for (let num of numSet) {
        // Initialize streak length for current number
        let length = 0;
        // Start with current number
        let current = num;
        
        // Keep squaring the number while it exists in our set
        while (numSet.has(current)) {
            length++;
            if (current > 100000) break; // Prevent overflow
            current = current * current;
        }
        
        // Only update maxLength if we found a streak of length > 1
        if (length > 1) {
            maxLength = Math.max(maxLength, length);
        }
    }
    
    // Return maxLength if we found a valid streak, otherwise return -1
    return maxLength > 1 ? maxLength : -1;
};

File: Solutions/2586-longest-square-streak-in-an-array/longest-square-streak-in-an-array.go
func longestSquareStreak(nums []int) int {
    // Create a map for O(1) lookup time
    numSet := make(map[int]bool)
    
    // Remove duplicates and add to map
    for _, num := range nums {
        numSet[num] = true
    }
    
    // Track the maximum streak length found
    maxLength := 0
    
    // Iterate through each number
    for num := range numSet {
        length := 0
        current := num
        
        // Keep squaring the number while it exists in our map
        for numSet[current] {
            length++
            // Check if squaring would overflow int
            if current > 46340 { // sqrt(MaxInt32) ≈ 46340
                break
            }
            current *= current
        }
        
        // Only update maxLength if we found a streak of length > 1
        if length > 1 {
            maxLength = max(maxLength, length)
        }
    }
    
    // Return maxLength if we found a valid streak, otherwise return -1
    if maxLength > 1 {
        return maxLength
    }
    return -1
}

// Helper function for max
func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}